endpoint.cpp

```diff
@@ -344,6 +344,48 @@ void Endpoint::print_statistics()
     printf("\n}\n");
 }
 
+void Endpoint::untrim_msg(const struct buffer *from, struct buffer *to)
```

Comment:
1. write the data you have
1. write the missing zeros
1. write the remaining data you need, if you do

instead of

1. copy everything to a new buffer, zero-extending it
1. write everything